sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala

```diff
@@ -259,4 +261,78 @@ class InsertIntoHiveTableSuite extends QueryTest with TestHiveSingleton with Bef
 
     sql("DROP TABLE table_with_partition")
   }
+
+  test("Reject partitioning that does not match table") {
+    sqlContext.setConf("hive.exec.dynamic.partition.mode", "nonstrict")
+    sql("CREATE TABLE partitioned (id bigint, data string) PARTITIONED BY (part string)")
+    val data = (1 to 10).map(i => (i, s"data-$i", if ((i % 2) == 0) "even" else "odd"))
+        .toDF("id", "data", "part")
+
+    intercept[AnalysisException] {
+      // cannot partition by 2 fields when there is only one in the table definition
+      data.write.partitionBy("part", "data").insertInto("partitioned")
+    }
+  }
+
+  test("Test partition mode = strict") {
+    sqlContext.conf.unsetConf("hive.exec.dynamic.partition.mode")
+    sql("CREATE TABLE partitioned (id bigint, data string) PARTITIONED BY (part string)")
+    val data = (1 to 10).map(i => (i, s"data-$i", if ((i % 2) == 0) "even" else "odd"))
+        .toDF("id", "data", "part")
+
+    intercept[SparkException] {
+      data.write.insertInto("partitioned")
+    }
+  }
+
+  test("Detect table partitioning") {
+    sqlContext.setConf("hive.exec.dynamic.partition.mode", "nonstrict")
+
+    sql("CREATE TABLE source (id bigint, data string, part string)")
+    val data = (1 to 10).map(i => (i, s"data-$i", if ((i % 2) == 0) "even" else "odd")).toDF()
+
+    data.write.insertInto("source")
+    checkAnswer(sql("SELECT * FROM source"), data.collect().toSeq)
+
+    sql("CREATE TABLE partitioned (id bigint, data string) PARTITIONED BY (part string)")
+    // this will pick up the output partitioning from the table definition
+    sqlContext.table("source").write.insertInto("partitioned")
+
+    checkAnswer(sql("SELECT * FROM partitioned"), data.collect().toSeq)
+
+    sqlContext.conf.unsetConf("hive.exec.dynamic.partition.mode")
+  }
+
+  test("Detect table partitioning with correct partition order") {
```

Comment:
@cloud-fan, this is the test that checks partition columns are in the correct order for the table (since Hive assumes they are when writing).